### PR TITLE
Comments flag filtering

### PR DESCRIPTION
**Pull request by jaleszek**
Comments flags sorting on the applications lists per Assessor/Admin.

### CODE REVIEW

File: app/search/form_answer_search.rb [hunk 1]
```diff
@@ -6,8 +6,22 @@ class FormAnswerSearch < Search
...
-    scoped_results.order(importance_flag: sort_order(!desc).to_sym).order(company_or_nominee_name: :asc)
+    section = (@subject.is_a?(Admin) ? "admin" : "critical")
+    section = Comment.sections[section]
+
+    scoped_results.select("form_answers.*, COUNT(comments.id) AS flags_count")
+      .joins("LEFT OUTER JOIN comments ON comments.commentable_id=form_answers.id")
+      .group("form_answers.id")
+      .where("(comments.section =? AND comments.commentable_type=? AND flagged =?)
+        OR comments.section IS NULL", section, "FormAnswer", true)
+      .order("flags_count #{sort_order(desc)}")
```
Comment (matthewford) [inline app/search/form_answer_search.rb]: Place the . on the previous line, together with the method call receiver.

File: app/search/form_answer_search.rb [hunk 2]
```diff
@@ -6,8 +6,22 @@ class FormAnswerSearch < Search
...
+
   def sort_by_flag(scoped_results, desc = false)
-    scoped_results.order(importance_flag: sort_order(!desc).to_sym).order(company_or_nominee_name: :asc)
+    section = (@subject.is_a?(Admin) ? "admin" : "critical")
+    section = Comment.sections[section]
+
+    scoped_results.select("form_answers.*, COUNT(comments.id) AS flags_count")
+      .joins("LEFT OUTER JOIN comments ON comments.commentable_id=form_answers.id")
+      .group("form_answers.id")
+      .where("(comments.section =? AND comments.commentable_type=? AND flagged =?)
```
Comment (matthewford) [inline app/search/form_answer_search.rb]: Place the . on the previous line, together with the method call receiver.

File: app/search/form_answer_search.rb [hunk 3]
```diff
@@ -6,8 +6,22 @@ class FormAnswerSearch < Search
...
+  end
+
   def sort_by_flag(scoped_results, desc = false)
-    scoped_results.order(importance_flag: sort_order(!desc).to_sym).order(company_or_nominee_name: :asc)
+    section = (@subject.is_a?(Admin) ? "admin" : "critical")
+    section = Comment.sections[section]
+
+    scoped_results.select("form_answers.*, COUNT(comments.id) AS flags_count")
+      .joins("LEFT OUTER JOIN comments ON comments.commentable_id=form_answers.id")
+      .group("form_answers.id")
```
Comment (matthewford) [inline app/search/form_answer_search.rb]: Place the . on the previous line, together with the method call receiver.

File: app/search/form_answer_search.rb [hunk 4]
```diff
@@ -6,8 +6,22 @@ class FormAnswerSearch < Search
...
+    super(scope)
+  end
+
   def sort_by_flag(scoped_results, desc = false)
-    scoped_results.order(importance_flag: sort_order(!desc).to_sym).order(company_or_nominee_name: :asc)
+    section = (@subject.is_a?(Admin) ? "admin" : "critical")
+    section = Comment.sections[section]
+
+    scoped_results.select("form_answers.*, COUNT(comments.id) AS flags_count")
+      .joins("LEFT OUTER JOIN comments ON comments.commentable_id=form_answers.id")
```
Comment (matthewford) [inline app/search/form_answer_search.rb]: Place the . on the previous line, together with the method call receiver.

**Comment by RuslanHamidullin** [comment]
Good to merge :+1:

File: app/search/form_answer_search.rb [hunk 5]
```diff
@@ -6,8 +6,22 @@ class FormAnswerSearch < Search
...
-    scoped_results.order(importance_flag: sort_order(!desc).to_sym).order(company_or_nominee_name: :asc)
+    section = (@subject.is_a?(Admin) ? "admin" : "critical")
+    section = Comment.sections[section]
+
+    scoped_results.select("form_answers.*, COUNT(comments.id) AS flags_count").
+      joins("LEFT OUTER JOIN comments on comments.commentable_id=form_answers.id").
+      group("form_answers.id").
+      where("(comments.section =? AND comments.commentable_type=? AND flagged =?)
+        OR comments.section IS NULL", section, "FormAnswer", true)
+      .order("flags_count #{sort_order(desc)}")
```
Comment (dreamfall) [inline app/search/form_answer_search.rb]: Please, for the readability prefer using AR API for SQL queries not raw SQL.

Comment (jaleszek) [inline app/search/form_answer_search.rb]: I'm using the AR API here, I just didn't wrap everything as I wouldn't like to introduce complexity related with arel methods (it won't be more readable then). If you see something obviously easy to refactor and improve the readability with reasonable cost let me know. Thanks

**Pull request merged**